training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala

```diff
@@ -0,0 +1,214 @@
+package com.airbnb.aerosolve.training
+
+import java.util
+
+import com.airbnb.aerosolve.core.models.BoostedStumpsModel
+import com.airbnb.aerosolve.core.models.DecisionTreeModel
+import com.airbnb.aerosolve.core.models.ForestModel
+import com.airbnb.aerosolve.core.Example
+import com.airbnb.aerosolve.core.ModelRecord
+import com.airbnb.aerosolve.core.util.Util
+import com.typesafe.config.Config
+import org.apache.spark.SparkContext
+import org.apache.spark.SparkContext._
+import org.apache.spark.rdd.RDD
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.util.Random
+import scala.util.Try
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+
+// A boosted tree forest trainer.
+// Alternates between fitting a tree and building one on the importance
+// sampled outliers of previous trees.
+object BoostedForestTrainer {
+  private final val log: Logger = LoggerFactory.getLogger("BoostedForestTrainer")
+  
+  case class BoostedForestTrainerOptions(candidateSize : Int,
+                                         rankKey : String,
+                                         rankThreshold : Double,
+                                         maxDepth : Int,
+                                         minLeafCount : Int,
+                                         numTries : Int,
+                                         splitCriteria : String,
+                                         numTrees :Int,
+                                         subsample : Double,
+                                         iterations : Int,
+                                         learningRate : Double)
+                                         
+  case class ForestResponse(tree : Int, leaf : Int, weight : Double)
+  case class ForestResult(label : Double, sum : Double, response : Array[ForestResponse])
+
+  def train(sc : SparkContext,
+            input : RDD[Example],
+            config : Config,
+            key : String) : ForestModel = {
+    val candidateSize : Int = config.getInt(key + ".num_candidates")
+    val rankKey : String = config.getString(key + ".rank_key")
+    val rankThreshold : Double = config.getDouble(key + ".rank_threshold")
+    val maxDepth : Int = config.getInt(key + ".max_depth")
+    val minLeafCount : Int = config.getInt(key + ".min_leaf_items")
+    val numTries : Int = config.getInt(key + ".num_tries")
+    val splitCriteria : String = Try(config.getString(key + ".split_criteria")).getOrElse("gini")
```

Comment:
A general comment not for this specific PR: we should refactor the config loading part to something like taskConfig = config.getConfig(key) then, param = taskConfig.getString(param_name).